Don't add implicits after non-given explicits.

mr-infty opened the pull request:
The function `TTImp.TTImp.implicitsAs` is changed so that implicit
arguments that come after explicit ones that are not given are no longer
added. This fixes #67.

The reasoning behind this is that the former behaviour---of adding all
missing implicits---makes `implicitsAs` return malformed terms in
certain cases where the underlying function type has implicit arguments
that come after explicit ones. For example, the interface declaration

    interface Foo a where
      bar : a -> {auto ok : ()} -> a

and the function declaration

    foo : Void -> {auto ok : ()} -> Void
    foo = ?foo_hole

result in the errors

    ok is not a valid argument in bar

and

    ok is not a valid argument in foo

respectively, because the auto-implicit is added during elaboration. For
instance, in the second example the left hand side of the pattern
matching declaration

    foo = ?foo_hole

gets replaced by `foo {ok=ok}` (in `implicitsAs`), which is an invalid
function call (according to `TTImp.Elab.App.checkAppWith`).

The changes made are as follows.

The functions `TTImp.TTImp.setAs` and `TTImp.TTImp.findImps` now both
take a new argument `es` of type `List (Maybe Name)`.

In `setAs`, this variable is used to record the explicit arguments which
are explicitly given in the left hand side of the pattern matching
declaration.

This variable is then passed onto `findImps`, which uses it to determine
when it encounters an explicit variable not given and stop searching for
implicits.

Russoul commented:
Just one note. It currently doesn't respect the "bind-all-explicits" pattern (it's new):
```
foo : a -> b -> c -> {auto x : a} -> a -- fails.
foo {} = x
```

`{}` binds however many explicits there is in the signature. So `x` should be implicitly bound on the LHS.

Just an FYI: `{}` are parsed as `UN "_"`. There is a brief description of it in the docs.

mr-infty commented:
> Just one note. It currently doesn't respect the "bind-all-explicits" pattern (it's new):
> 
> ```
> foo : a -> b -> c -> {auto x : a} -> a -- fails.
> foo {} = x
> ```
> 
> `{}` binds however many explicits there is in the signature. So `x` should be implicitly bound on the LHS.
> 
> Just an FYI: `{}` are parsed as `UN "_"`. There is a brief description of it in the docs.

Thanks for the heads up! I've read the documentation, but just to make sure I understand it correctly: is the underscore notation supposed to skip *all* explicit arguments (as suggested by the example `dontCare`) or only all *named* explicit arguments (as suggested by the last sentence before the examples)?

Russoul commented:
It always skips both named and unnamed explicits regardless the form: `{}` or `{x, y, z, ... , _}`

Russoul commented:
Yep, that last phrase in the docs before the examples is misleading.

mr-infty commented:
@Russoul 
I've updated the code to make it compatible with the new syntax. I've also added your counterexample as a test-case in the regression test.

gallais commented:
> Just an FYI: `{}` are parsed as `UN "_"`.

This seems brittle. Is there a deep reason for not adding a new constructor
to the name datatype for `{}`?

Russoul commented:
I don't think there is a reason at all. Maybe only that `Name` is overused throughout Idris 2. Even core TT (`Term`) is indexed over `List Name`. While in fact, only variables, i.e `UN` and `MN`, are valid in `Term`. 
And this is what I am working on right now: replacing `Name` to a more precise datatype when appropriate in Idris codebase.
I'll remember to throw away `UN _` and have a constructor in `Name` for the bind-all-explicits pattern.

gallais commented on src/TTImp/TTImp.idr:
```diff
@@ -496,26 +496,26 @@ findImplicits tm = []
-    setAs : List (Maybe Name) -> RawImp -> Core RawImp
-    setAs is (IApp loc f a)
-        = do f' <- setAs is f
+    setAs : List (Maybe Name) -> List (Maybe Name) -> RawImp -> Core RawImp
```
Can we get a comment documenting the different roles of these two
`List (Maybe Name)` arguments?

mr-infty replied:
Sure, I'll add a comment in order to make it clearer to the reader!

gallais commented on src/TTImp/TTImp.idr:
```diff
@@ -529,20 +529,27 @@ implicitsAs defs ns tm = setAs (map Just (ns ++ map UN (findIBinds tm))) tm
-        findImps : List (Maybe Name) -> NF [] -> Core (List (Name, PiInfo RawImp))
-        findImps ns (NBind fc x (Pi _ _ Explicit _) sc)
-            = findImps ns !(sc defs (toClosure defaultOpts [] (Erased fc False)))
+        findImps : List (Maybe Name) -> List (Maybe Name) -> NF [] -> Core (List (Name, PiInfo RawImp))
```
Same comment here: a bit of documentation will make it easier for
future readers to understand what is going on.

gallais commented on tests/idris2/reg036/Test.idr:
```diff
@@ -0,0 +1,10 @@
+foo = ?foo_hole
+
+baz : a -> b -> c -> {auto x : a} -> a
+baz {} = x
```
I think I have found a bug involving shadowing:

```idris
foo : {a : Type} -> Void -> {a : a} -> Void -> {auto ok : ()} -> Void
foo _ {a = c} = ?foo_hole
-- ^   ^--- This should bind the second implicit `a`,
-- |        the one that comes after the first explicit argument
-- `-- there should be an implicitly bound `a : Type` here
```

gives me a hole with `c : Type` rather than `c : a`.

mr-infty replied:
I'm surprised to learn that this is legal Idris code. Does Idris have separate namespaces for term and type variables?

gallais replied:
Why shouldn't it be legal? It's not a namespace issue, it's classic
shadowing in a telescope: when checking the type `a` in the telescope
`pre -> (x : a) -> post`, the variables bound in `pre` are in scope
(but not `x` itself as no binding is self-referential) and when checking
`post`, the variables `pre, x` are in scope (thus any `x` in `pre` is
shadowed by the most-recently bound one).

mr-infty replied:
Ah, I see. Thanks for clarifying that for me!

gallais reviewed (commented):
I think I have found a bug involving shadowing.

gallais commented:
I guess the bug I have found precedes this PR because `master` currently fails
in a similar way with:

```idris
foo : {a : Type} -> Void -> {a : a} -> Void
foo _ {a = c} = ?foo_hole
```

So I'd be willing to merge this PR with the only change being the added documentation.

mr-infty commented:
@gallais I've added some documentation. I hope, it makes the meaning of the code clearer.

gallais commented:
Thank you, this is very good. It has even clarified some of the things
I had not realised after carefully reading this PR. Looks to me like
the new application syntax is quite ambiguous. I'll open another issue.
